# OLS Linear Regression
OLS: ordinary least squares

## Example 1: does year of education increase income?
In the nls97 dataset, there is a variable called `highestgradecompleted.` It measures the highest grade of regular school the respondent has completed, typically ranging from 0 (no formal schooling) to 20 (completion of 8 or more years of higher education).

Regression equation:
$I_i = a + bE_i + \mu$, where $I_i$ represents income of individual $i$, and $E_i$ represents the years of formal school education.
Hypotheses:
- $H_0$: $b=0$
- $H_1$: $b \neq 0$

In [12]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.api as sm

df = pd.read_csv("../05-data_cleaning/data/nls97.csv")
df.head()

,personid,gender,birthmonth,birthyear,highestgradecompleted,maritalstatus,childathome,childnotathome,wageincome,weeklyhrscomputer,...,colenrfeb13,colenroct13,colenrfeb14,colenroct14,colenrfeb15,colenroct15,colenrfeb16,colenroct16,colenrfeb17,colenroct17
0,100061,Female,5,1980,13.0,Married,4.0,0.0,12500.0,10 hours or more a week,...,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled
1,100139,Male,9,1983,12.0,Married,2.0,0.0,120000.0,1 to 3 hours a week,...,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled
2,100284,Male,11,1984,7.0,Never-married,1.0,0.0,58000.0,NaN,...,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled
3,100292,Male,4,1982,NaN,NaN,NaN,NaN,NaN,NaN,...,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,NaN,NaN,NaN,NaN
4,100583,Male,1,1980,13.0,Married,4.0,0.0,30000.0,Less than 1 hour a week,...,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled,1. Not enrolled


### Prepare data

In [ ]:
df['school_year'] = pd.to_numeric(df['highestgradecompleted'], errors='coerce')

df.dropna(subset=['school_year', 'wageincome'], inplace=True)

### Run regression


In [8]:

X = df["school_year"]
y = df["wageincome"]

# Add constant (intercept)
X = sm.add_constant(X)

model = sm.OLS(y, X).fit()

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             wageincome   R-squared:                       0.063
Model:                            OLS   Adj. R-squared:                  0.062
Method:                 Least Squares   F-statistic:                     336.9
Date:                Tue, 31 Mar 2026   Prob (F-statistic):           6.85e-73
Time:                        18:38:09   Log-Likelihood:                -60623.
No. Observations:                5051   AIC:                         1.212e+05
Df Residuals:                    5049   BIC:                         1.213e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        1.358e+04   2041.940      6.652      

### Interpretation
- $a=13580$, indicating people with no formal school year education still have income on average of \\$13,580.  This does not mean everyone who does not have education still earn \$13,580, it simply is the model intercept. (I usually try not to interpret the value of a).
- $b=2497.20$, indicating each extra year of school education increases income by \$2,497.
- $p < 0.001$ (in fact, p value can be access using `model.pvalues` and $p=6.85 \times 10^{-73}$, extremely small, so the coefficient $b$ is statistically significant.
- R-square: Proportion of variation in income explained by education

In [18]:
alpha = 0.05

if model.pvalues["school_year"] < alpha:
    print("Reject H0: education significantly affects income")
else:
    print("Fail to reject H0")

Reject H0: education significantly affects income


### Vie